cairo-libs/packages/cometbft/src/utils.cairo

```diff
@@ -33,3 +33,69 @@ impl FractionAsProtoName of ProtoName<Fraction> {
         "Fraction"
     }
 }
+
+pub impl SpanU8TryIntoU256 of TryInto<Span<u8>, u256> {
+    /// Decodes as big endian.
+    fn try_into(self: Span<u8>) -> Option<u256> {
+        // Only allows size 32 to ensure all bytes fit exactly into a `u256`.
+        if (self.len() != 32) {
+            return Option::None;
+        }
+
+        // Precomputed powers to remove the overhead of repeated multiplications.
+        let mut ret: u256 = 0;
+        let two_pow_0 = 0x1;
+        let two_pow_1 = 0x100;
+        let two_pow_2 = 0x10000;
+        let two_pow_3 = 0x1000000;
+        let two_pow_4 = 0x100000000;
+        let two_pow_5 = 0x10000000000;
+        let two_pow_6 = 0x1000000000000;
+        let two_pow_7 = 0x100000000000000;
+        let two_pow_8 = 0x10000000000000000;
+        let two_pow_9 = 0x1000000000000000000;
+        let two_pow_10 = 0x100000000000000000000;
+        let two_pow_11 = 0x10000000000000000000000;
+        let two_pow_12 = 0x1000000000000000000000000;
+        let two_pow_13 = 0x100000000000000000000000000;
+        let two_pow_14 = 0x10000000000000000000000000000;
+        let two_pow_15 = 0x1000000000000000000000000000000;
+
+        // No loop overhead with manual unrolling.
+        ret.low += (*self[31]).into() * two_pow_0;
+        ret.low += (*self[30]).into() * two_pow_1;
+        ret.low += (*self[29]).into() * two_pow_2;
+        ret.low += (*self[28]).into() * two_pow_3;
+        ret.low += (*self[27]).into() * two_pow_4;
+        ret.low += (*self[26]).into() * two_pow_5;
+        ret.low += (*self[25]).into() * two_pow_6;
+        ret.low += (*self[24]).into() * two_pow_7;
+        ret.low += (*self[23]).into() * two_pow_8;
+        ret.low += (*self[22]).into() * two_pow_9;
+        ret.low += (*self[21]).into() * two_pow_10;
+        ret.low += (*self[20]).into() * two_pow_11;
+        ret.low += (*self[19]).into() * two_pow_12;
+        ret.low += (*self[18]).into() * two_pow_13;
+        ret.low += (*self[17]).into() * two_pow_14;
+        ret.low += (*self[16]).into() * two_pow_15;
```

Comment:
actually, we don't really need `two_pow_*`. you can just do this.

```cairo
let mut ret = 0_u256;

// ret.low == 0 
ret.low = ret.low * 0x100 + (*self[16]).into(); // ret.low << 8 | self[16]
ret.low = ret.low * 0x100 + (*self[17]).into(); // ret.low << 8 | self[17]
ret.low = ret.low * 0x100 + (*self[18]).into(); // ret.low << 8 | self[18]
// ...
ret.low = ret.low * 0x100 + (*self[32]).into(); // ret.low << 8 | self[32]

// same for ret.hight
```

Reply:
ah sorry. it should be `31` -- I mistyped as `32` (off by one 😅)

here is an example working code in Rust

https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=c43817f288dbaf44056625bcda2fe4fc

Reply:
Oh, nice. Got it. Ended up with the same gas, but less verbose. Thanks.
[d935c70](https://github.com/informalsystems/ibc-starknet/pull/361/commits/d935c701f91f12cf0f4ea8066296a89d28938fc2)